gpytorch/lazy/kronecker_product_added_diag_lazy_tensor.py

```diff
@@ -115,36 +163,92 @@ def _solve(self, rhs, preconditioner=None, num_tridiag=0):
+                    )
+                    evals, evecs = symm_prod.symeig(eigenvectors = True)
+                    evals_p_i = DiagLazyTensor(evals + 1.)
+                    res = dlt_inv_root.matmul(
+                        evecs.matmul(
+                            evals_p_i.inv_matmul(
+                                evecs._transpose_nonbatch().matmul(dlt_inv_root.matmul(rhs))))
+                    )
+                    return res.to(rhs_dtype)
 
             tmp_term = S.matmul(Lambda_I.inv_matmul(S._transpose_nonbatch().matmul(rhs)))
-            res = lt._inv_matmul(rhs - tmp_term)  # performed efficiently b/c lt has Kronecker structure
+
+            # TODO: check why these are different!
+            if isinstance(lt, KroneckerProductLazyTensor):
+                res = lt._inv_matmul(rhs - tmp_term)
+            elif isinstance(lt, KroneckerProductAddedDiagLazyTensor):
+                res = lt._solve(rhs - tmp_term)
+
             return res.to(rhs_dtype)
 
         # in all other cases we fall back to the default
-        super()._solve(rhs, preconditioner=preconditioner, num_tridiag=num_tridiag)
+        return super()._solve(rhs, preconditioner=preconditioner, num_tridiag=num_tridiag)
 
     def _root_decomposition(self):
         if self._diag_is_constant:
             evals, q_matrix = self.lazy_tensor.symeig(eigenvectors=True)
             updated_evals = DiagLazyTensor((evals + self.diag_tensor.diag()).pow(0.5))
             return MatmulLazyTensor(q_matrix, updated_evals)
-        super()._root_decomposition()
+        return super()._root_decomposition()
 
     def _root_inv_decomposition(self, initial_vectors=None):
         if self._diag_is_constant:
             evals, q_matrix = self.lazy_tensor.symeig(eigenvectors=True)
             inv_sqrt_evals = DiagLazyTensor((evals + self.diag_tensor.diag()).pow(-0.5))
             return MatmulLazyTensor(q_matrix, inv_sqrt_evals)
-        super()._root_inv_decomposition(initial_vectors=initial_vectors)
+        return super()._root_inv_decomposition(initial_vectors=initial_vectors)
+
+    def _symeig(self, eigenvectors: bool = False) -> Tuple[Tensor, Optional[LazyTensor]]:
+        # return_evals_as_lazy is a flag to return the eigenvalues as a lazy tensor
+        # which is useful for root decompositions here (see the root_decomposition
+        # method above)
+        if self._diag_is_constant:
+            evals, evecs = self.lazy_tensor.symeig(eigenvectors=eigenvectors)
+            evals = evals + self.diag_tensor.diag_values
+
+            return evals, evecs
+        return super()._symeig(eigenvectors=eigenvectors)
+
+    def __add__(self, other):
+        if isinstance(other, ConstantDiagLazyTensor):
+            if not self._diag_is_constant:
+                new_kpadlt = KroneckerProductAddedDiagLazyTensor(self.lazy_tensor, other)
+                return KroneckerProductAddedDiagLazyTensor(new_kpadlt, self.diag_tensor)
```

Comment:
should we instead return sth like

```
new_diag = DiagLazyTensor(self.diag_tensor.diag() + other.diag())
return KroneckerProductAddedDiagLazyTensor(self.lazy_tensor, new_diag)
```

Reply:
The goal was to create a special returning of KPAddedDiagLT for the (K + a I) + D case; but, supporting that isn't a great idea. Should be resolved now.